### PR TITLE
[FLOC-1755] Don't mess with in use datasets in block device deployer

### PR DESCRIPTION
**Pull request by itamarst**

### CODE REVIEW

File: flocker/node/_deploy.py [hunk 1]
```diff
@@ -435,6 +435,41 @@ def run(self, deployer):
         return gather_deferreds(results)
 
 
+class NotInUseDatasets(object):
+    """
+    Filter out datasets that are in use by applications.
+
+    For now we delay things like deletion we know applications
+    aren't using the dataset. Later on we'll use leases to decouple
```
Comment (exarkun) [inline flocker/node/_deploy.py]: Missing a word here, perhaps?  I don't quite follow this sentence.

File: flocker/node/_deploy.py [hunk 2]
```diff
@@ -435,6 +435,41 @@ def run(self, deployer):
...
+    the application and dataset logic better.
+    """
+    def __init__(self, node_state):
```
Comment (exarkun) [inline flocker/node/_deploy.py]: Maybe a class method, `from_node_state(node_state)`

File: flocker/node/agents/blockdevice.py [hunk 3]
```diff
@@ -1511,6 +1513,16 @@ def _mountpath_for_dataset_id(self, dataset_id):
     def calculate_changes(self, configuration, cluster_state):
         this_node_config = configuration.get_node(
             self.node_uuid, hostname=self.hostname)
+        local_state = cluster_state.get_node(self.node_uuid,
+                                             hostname=self.hostname)
+
+        # We need to know applications (for now) to see if we should delay
+        # deletion or handoffs. Eventually this will rely on leases instead.
+        if local_state.applications is None:
+            return in_parallel(changes=[])
```
Comment (exarkun) [inline flocker/node/agents/blockdevice.py]: Was there another issue filed for this part?  It's pretty related, so maybe not... But it'd be nice to make sure it gets closed at the same time if so.

Comment (exarkun) [inline flocker/node/agents/blockdevice.py]: I guess there should be a link to an issue about leases here, too.

File: flocker/node/agents/test/test_blockdevice.py [hunk 4]
```diff
@@ -588,6 +590,24 @@ class ScenarioMixin(object):
         }
     )
 
+    @staticmethod
+    def add_application_with_volume(node_state):
```
Comment (exarkun) [inline flocker/node/agents/test/test_blockdevice.py]: A free function, instead?

**Comment by exarkun** [comment]
Thanks.  This looks good.  Address minor inline comments and merge if the buildbot failures are unrelated.

File: flocker/node/_deploy.py (hunk 2 above)
Comment (itamarst) [inline flocker/node/_deploy.py]: Well, the implementation is going to change anyway when we switch to leases, so having a specific `__init__`? Eh. If this ever becomes more generally useful maybe.

File: flocker/node/agents/blockdevice.py (hunk 3 above)
Comment (itamarst) [inline flocker/node/agents/blockdevice.py]: I don't think there was another issue, no.

**Pull request merged**